File: test/mono_euroc.cpp
/**
 * EuRoC dataset test
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <unistd.h>

#include <opencv2/opencv.hpp>

#include "ov2slam.hpp"
#include "visualize.hpp"

void load_images(const std::string& img_folder, const std::string &img_ts_path, 
  std::vector<std::string>& img_path, std::vector<double>& img_ts) 
{
  std::ifstream ifs;
  ifs.open(img_ts_path.c_str());

  if( !ifs.is_open() ) {
    std::cerr << "Cannot open file: " << img_ts_path << std::endl;
    exit(-1);
  }

  img_path.reserve(5000);
  img_ts.reserve(5000);

  while( !ifs.eof() ) {
    std::string s;
    getline(ifs, s);
    if( !s.empty() ){
      std::stringstream ss;
      ss << s;
      img_path.push_back(img_folder + "/" + ss.str() + ".png");
      double t;
      ss >> t;
      img_ts.push_back(t / 1e9);
    }
  }
}

// ./mono_euroc ../parameters_files/fast/euroc/euroc_mono.yaml ~/dataset/mav0/cam0/data ../example/euroc_ts/V101.txt

int main(int argc, char** argv) {
    std::vector<std::string> img_path;
    std::vector<double> img_ts;

    load_images(std::string(argv[2]), std::string(argv[3]), img_path, img_ts);

    // Load the parameters
    std::string parameters_file = argv[1];

    const cv::FileStorage fsSettings(parameters_file.c_str(), cv::FileStorage::READ);
    if(!fsSettings.isOpened()) {
        std::cout << "Failed to open settings file";
        return 1;
    }

    std::shared_ptr<SlamParams> pparams;
    pparams.reset( new SlamParams(fsSettings) );

    std::shared_ptr<SlamManager> pslam;
    pslam.reset( new SlamManager(pparams) );

    std::shared_ptr<Visualize> pviz;
    pviz.reset( new Visualize(pslam) );

    std::thread slam_thread(&SlamManager::run, pslam);
    slam_thread.detach();
    
    std::thread viz_thread(&Visualize::run, pviz);
    viz_thread.detach();

    for(size_t i = 0; i < img_path.size(); ++i) {
        cv::Mat im = cv::imread(img_path[i], CV_LOAD_IMAGE_UNCHANGED);
        pslam->addNewMonoImage(img_ts[i], im);
        usleep(1000 * 1000 / 30);
    }

  return 0;
}

File: test/mono_euroc.py
import cv2
import sys
import ov2slam
import time


def load_image(img_folder, img_ts_path):
  imgs = []
  ts = []
  
  with open(img_ts_path, 'r') as f:
    ts = f.read().strip().split()
  
  for(_, t) in enumerate(ts):
    img_path = img_folder + '/' + t + '.png'
    imgs.append(img_path)
  
  return imgs, ts

# python mono_euroc.py parameters_files/accurate/euroc/euroc_mono.yaml ~/dataset/mav0/cam0/data euroc_ts/V101.txt

if __name__ == '__main__' :
  conf, img_folder, img_ts = sys.argv[1], sys.argv[2], sys.argv[3]
  
  imgs, ts = load_image(img_folder, img_ts)
  
  session = ov2slam.Session.create(conf)
  session.startVisualize()
  
  input('wait for pangolin')
  
  for i in range(0, len(imgs)):
    img = cv2.imread(imgs[i])
    session.addTrack(img, float(ts[i]))
    time.sleep(1/30)
    
  input()
  session.stop()